ethereum/spec/src/main/java/tech/pegasys/teku/spec/datastructures/blobs/versions/deneb/SignedBlobSidecarsUnblinderDeneb.java

```diff
@@ -0,0 +1,85 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.spec.datastructures.blobs.versions.deneb;
+
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.spec.datastructures.blobs.SignedBlobSidecarsUnblinder;
+import tech.pegasys.teku.spec.schemas.SchemaDefinitionsDeneb;
+
+public class SignedBlobSidecarsUnblinderDeneb implements SignedBlobSidecarsUnblinder {
+
+  private final BlobSidecarSchema blobSidecarSchema;
+  private final SignedBlobSidecarSchema signedBlobSidecarSchema;
+  private final List<SignedBlindedBlobSidecar> signedBlindedBlobSidecars;
+
+  private volatile SafeFuture<BlobsBundle> blobsBundleFuture;
+
+  public SignedBlobSidecarsUnblinderDeneb(
+      final SchemaDefinitionsDeneb schemaDefinitions,
+      final List<SignedBlindedBlobSidecar> signedBlindedBlobSidecars) {
+    this.blobSidecarSchema = schemaDefinitions.getBlobSidecarSchema();
+    this.signedBlobSidecarSchema = schemaDefinitions.getSignedBlobSidecarSchema();
+    this.signedBlindedBlobSidecars = signedBlindedBlobSidecars;
+  }
+
+  @Override
+  public void setBlobsBundleSupplier(final Supplier<SafeFuture<BlobsBundle>> blobsBundleSupplier) {
+    blobsBundleFuture = blobsBundleSupplier.get();
+  }
+
+  @Override
+  public SafeFuture<List<SignedBlobSidecar>> unblind() {
+    return blobsBundleFuture.thenApply(
+        blobsBundle ->
+            signedBlindedBlobSidecars.stream()
+                .map(
+                    signedBlindedBlobSidecar ->
+                        unblindSignedBlindedBlobSidecar(signedBlindedBlobSidecar, blobsBundle))
+                .collect(Collectors.toUnmodifiableList()));
+  }
+
+  private SignedBlobSidecar unblindSignedBlindedBlobSidecar(
+      final SignedBlindedBlobSidecar signedBlindedBlobSidecar, final BlobsBundle blobsBundle) {
+    final BlindedBlobSidecar blindedBlobSidecar = signedBlindedBlobSidecar.getBlindedBlobSidecar();
+    final Blob blob = findBlobWithRoot(blobsBundle, blindedBlobSidecar.getBlobRoot());
+    final BlobSidecar unblindedBlobSidecar =
+        blobSidecarSchema.create(
+            blindedBlobSidecar.getBlockRoot(),
+            blindedBlobSidecar.getIndex(),
+            blindedBlobSidecar.getSlot(),
+            blindedBlobSidecar.getBlockParentRoot(),
+            blindedBlobSidecar.getProposerIndex(),
+            blob,
+            blindedBlobSidecar.getKZGCommitment(),
+            blindedBlobSidecar.getKZGProof());
+    return signedBlobSidecarSchema.create(
+        unblindedBlobSidecar, signedBlindedBlobSidecar.getSignature());
+  }
+
+  private Blob findBlobWithRoot(final BlobsBundle blobsBundle, final Bytes32 blobRoot) {
```

Comment:
I'm not super convinced here. I know that this is safer but shouldn't we have bundle and blindedSidecars aligned? so we can lookup by index instead of searching a matching root? We could lookup by index and then check that roots corresponds. If index is out of bounds then it's a sign of misalignment

Reply:
yeah I like that approach more, fixed it